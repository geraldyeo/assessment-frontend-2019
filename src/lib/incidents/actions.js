export const CREATE_ONE_INCIDENT = 'CREATE_ONE_INCIDENT';

export const createOneIncident = (payload) => ({ type: CREATE_ONE_INCIDENT, payload });