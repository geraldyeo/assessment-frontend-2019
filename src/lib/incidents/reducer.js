import React, { createContext, useReducer } from 'react';
import { CREATE_ONE_INCIDENT } from './actions';

export const initialState = {
  incidents: [
    { key: 'bcg_1', title: 'First incident', assignee: 'Admin', status: 'Resolved' }
  ]
};

export const reducer = (state, { type, payload }) => {
  Object.freeze(state);

  /* eslint-disable no-fallthrough */
  switch(type) {
    case CREATE_ONE_INCIDENT: {
      if (payload) {
        console.log(payload);
        return {...state, incidents: [...state.incidents, payload]};
      }
    }
    default:
      return state;
  }
  /* eslint-enable no-fallthrough */
};

  export const IncidentContext = createContext(null);

  export const IncidentProvider = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    return (
      <IncidentContext.Provider value={{ state, dispatch }}>
        {children}
      </IncidentContext.Provider>
    );
  }