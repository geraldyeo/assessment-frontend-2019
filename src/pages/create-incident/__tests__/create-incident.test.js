import React from 'react';
import { fireEvent, render, cleanup, waitForElement, waitForDomChange, act } from '@testing-library/react';
import 'jest-dom/extend-expect';
import { CreateIncident } from '../index';
import { IncidentProvider } from '../../../lib/incidents/reducer'



afterEach(cleanup);

describe('CreateIncident component', () => {
  test('submitting empty form should display errors', async () => {
    const {getByTestId, getAllByTestId, container} = render(<CreateIncident />, {wrapper: IncidentProvider});
    // check all form inputs are there
    expect(getByTestId('create-incident-container')).toBeDefined();
    expect(getByTestId('title-inp')).toBeDefined();
    expect(getByTestId('assignee-inp')).toBeDefined();
    expect(getByTestId('status-cb')).toBeDefined();
    expect(getByTestId('submit-incident')).toBeDefined();
    // submit empty form
    fireEvent.click(getByTestId('submit-incident'));
    const errorMessages = await waitForElement(() => getAllByTestId('error-message'), {container});
    expect(errorMessages).toHaveLength(2);
  });

  test('checking resolve checkbox should change card color', async () => {
    const {getByTestId} = render(<CreateIncident />, {wrapper: IncidentProvider});
    const incident = getByTestId('create-incident-container');
    // initial bgcolor
    expect(window.getComputedStyle(incident).backgroundColor).toMatchSnapshot();
    fireEvent.click(getByTestId('status-cb'));

    expect(window.getComputedStyle(incident).backgroundColor).toMatchSnapshot();
  });

  test('submitting completed form should reset form', async () => {
    const {getByTestId, container} = render(<CreateIncident />, {wrapper: IncidentProvider});
    const titleInp = getByTestId('title-inp');
    const assigneeInp = getByTestId('assignee-inp');
    const submitBtn = getByTestId('submit-incident');

    fireEvent.change(titleInp, {target:{value:'Second Event'}});
    expect(titleInp.value).toBe('Second Event');

    fireEvent.change(assigneeInp, {target:{value:'User'}});
    expect(assigneeInp.value).toBe('User');

    fireEvent.click(submitBtn);
    const mutated = await waitForDomChange({container});
    expect(mutated).toHaveLength(2);
    expect(titleInp.value).toBe('');
    expect(assigneeInp.value).toBe('');
  });
});