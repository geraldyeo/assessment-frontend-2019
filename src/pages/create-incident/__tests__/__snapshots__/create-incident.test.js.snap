// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateIncident component checking resolve checkbox should change card color 1`] = `"rgb(251, 244, 227)"`;

exports[`CreateIncident component checking resolve checkbox should change card color 2`] = `"rgb(216, 237, 234)"`;
